cli/cache/http_cache/local.rs

```diff
@@ -0,0 +1,564 @@
+
+  fn set(
+    &self,
+    url: &Url,
+    headers: crate::http_util::HeadersMap,
+    content: &[u8],
+  ) -> Result<(), AnyError> {
+    let is_redirect = headers.contains_key("location");
+    if !is_redirect {
+      let cache_filepath = self.get_cache_filepath(url, &headers)?;
+      // Create parent directory
+      let parent_filename = cache_filepath.parent().unwrap();
+      ensure_dir_exists(parent_filename)?;
+      // Cache content
+      atomic_write_file(&cache_filepath, content, CACHE_PERM).with_context(
+        || format!("Failed saving '{}'", cache_filepath.display()),
+      )?;
+    }
+
+    let sub_path = url_to_local_sub_path(url, &headers)?;
+    self.manifest.insert_data(sub_path, url.clone(), headers);
+
+    Ok(())
+  }
+
+  fn read_file_bytes(
+    &self,
+    key: &HttpCacheItemKey,
+  ) -> Result<Option<Vec<u8>>, AnyError> {
+    debug_assert!(key.is_local_key);
+
+    let metadata = self.get_url_metadata_checking_global_cache(key.url)?;
+    match metadata {
+      Some(data) => {
+        if data.is_redirect() {
+          // return back an empty file for redirect
+          Ok(Some(Vec::new()))
+        } else {
+          // if it's not a redirect, then it should have a file path
+          let cache_filepath =
+            self.get_cache_filepath(key.url, &data.headers)?;
+          Ok(read_file_bytes(&cache_filepath)?)
+        }
+      }
+      None => Ok(None),
+    }
+  }
+
+  fn read_metadata(
+    &self,
+    key: &HttpCacheItemKey,
+  ) -> Result<Option<CachedUrlMetadata>, AnyError> {
+    debug_assert!(key.is_local_key);
+
+    self.get_url_metadata_checking_global_cache(key.url)
+  }
+}
+
+struct LocalCacheSubPath {
+  pub has_hash: bool,
+  pub parts: Vec<String>,
+}
+
+impl LocalCacheSubPath {
+  pub fn as_path_from_root(&self, root_path: &Path) -> PathBuf {
+    let mut path = root_path.to_path_buf();
+    for part in &self.parts {
+      path.push(part);
+    }
+    path
+  }
+}
+
+fn url_to_local_sub_path(
+  url: &Url,
+  headers: &HeadersMap,
+) -> Result<LocalCacheSubPath, UrlToFilenameConversionError> {
+  // https://stackoverflow.com/a/31976060/188246
+  static FORBIDDEN_CHARS: Lazy<HashSet<char>> = Lazy::new(|| {
+    HashSet::from(['?', '<', '>', ':', '*', '|', '\\', ':', '"', '\'', '/'])
+  });
+
+  fn has_forbidden_chars(segment: &str) -> bool {
+    segment.chars().any(|c| {
+      let is_uppercase = c.is_ascii_alphabetic() && !c.is_ascii_lowercase();
+      FORBIDDEN_CHARS.contains(&c)
+        // do not allow uppercase letters in order to make this work
+        // the same on case insensitive file systems
+        || is_uppercase
+    })
+  }
+
+  fn has_known_extension(path: &str) -> bool {
+    let path = path.to_lowercase();
+    path.ends_with(".js")
+      || path.ends_with(".ts")
+      || path.ends_with(".jsx")
+      || path.ends_with(".tsx")
+      || path.ends_with(".mts")
+      || path.ends_with(".mjs")
+      || path.ends_with(".json")
+      || path.ends_with(".wasm")
```

Comment:
Just to double check - we don't need to care about cjs files here?

Reply:
I've been leaving cjs/cts files out of a lot of places to discourage their use https://github.com/denoland/deno/issues/17650